github/repos_rules.go

```diff
@@ -0,0 +1,392 @@
+type RulesetLinks struct {
+	Self *RulesetLink `json:"self,omitempty"`
+}
+
+// RulesetRefConditionParameters represents the conditions object for ref_names.
+type RulesetRefConditionParameters struct {
+	Include []string `json:"include"`
+	Exclude []string `json:"exclude"`
+}
+
+// RulesetRepositoryConditionParameters represents the conditions object for repository_names.
+type RulesetRepositoryConditionParameters struct {
+	Include   []string `json:"include"`
+	Exclude   []string `json:"exclude"`
+	Protected *bool    `json:"protected,omitempty"`
+}
+
+// RulesetCondition represents the conditions object in a ruleset.
+type RulesetConditions struct {
+	RefName        *RulesetRefConditionParameters        `json:"ref_name,omitempty"`
+	RepositoryName *RulesetRepositoryConditionParameters `json:"repository_name,omitempty"`
+}
+
+// RulePatternParameters represents the rule pattern parameters.
+type RulePatternParameters struct {
+	Name *string `json:"name,omitempty"`
+	// If Negate is true, the rule will fail if the pattern matches.
+	Negate *bool `json:"negate,omitempty"`
+	// Possible values for Operator are: starts_with, ends_with, contains, regex
+	Operator string `json:"operator"`
+	Pattern  string `json:"pattern"`
+}
+
+// UpdateAllowsFetchAndMergeRuleParameters represents the update rule parameters.
+type UpdateAllowsFetchAndMergeRuleParameters struct {
+	UpdateAllowsFetchAndMerge bool `json:"update_allows_fetch_and_merge"`
+}
+
+// RequiredDeploymentEnvironmentsRuleParameters represents the required_deployments rule parameters.
+type RequiredDeploymentEnvironmentsRuleParameters struct {
+	RequiredDeploymentEnvironments []string `json:"required_deployment_environments"`
+}
+
+// PullRequestRuleParameters represents the pull_request rule parameters.
+type PullRequestRuleParameters struct {
+	DismissStaleReviewsOnPush      bool `json:"dismiss_stale_reviews_on_push"`
+	RequireCodeOwnerReview         bool `json:"require_code_owner_review"`
+	RequireLastPushApproval        bool `json:"require_last_push_approval"`
+	RequiredApprovingReviewCount   int  `json:"required_approving_review_count"`
+	RequiredReviewThreadResolution bool `json:"required_review_thread_resolution"`
+}
+
+// RuleRequiredStatusChecks represents the RequiredStatusChecks for the RequiredStatusChecksRuleParameters object.
+type RuleRequiredStatusChecks struct {
+	Context       string `json:"context"`
+	IntegrationID *int64 `json:"integration_id,omitempty"`
+}
+
+// RequiredStatusChecksRuleParameters represents the required_status_checks rule parameters.
+type RequiredStatusChecksRuleParameters struct {
+	RequiredStatusChecks             []RuleRequiredStatusChecks `json:"required_status_checks"`
+	StrictRequiredStatusChecksPolicy bool                       `json:"strict_required_status_checks_policy"`
+}
+
+// RepositoryRule represents a GitHub Rule.
+type RepositoryRule struct {
+	Type       string      `json:"type"`
+	Parameters interface{} `json:"parameters,omitempty"`
```

Comment:
Instead of `interface{}`, I'm wondering if this would be better to have a `*json.RawMessage` here like we do in other places in this repo?

Reply:
Can I get some guidance on how this might be done? I took a look at some of the other examples and they seem to suggest to allow users to pass in `*json.RawMessage` types as the parameters? I am on the fence with this one as I feel like you lose out on all the typed constructs when you switch over to this it in this particular case?

```
# Example of creating a pull request rule
func NewPullRequestRule(params *json.RawMessage) (rule *RepositoryRule) {
	return &RepositoryRule{
		Type:       "pull_request",
		Parameters: params,
	}
}

pullRequestParams := json.RawMessage(`{
	"dismiss_stale_reviews_on_push": true,
	"require_code_owner_review": true,
	"require_last_push_approval": true,
	"required_approving_review_count": 1,
	"required_review_thread_resolution": true
}`)

rule := NewPullRequestRule(&pullRequestParams),
```

Reply:
Sorry, I'm OOO today and will take a look tomorrow.

Reply:
In general, you don't have to give up strong typing with `json.RawMessage`... you can always marshal other strongly-typed structs into the raw message.  It is just another way to handle variadic sub-messages within a JSON blob.
So I'm not sure you lose any functionality with `json.RawMessage` over the use of `any`... I think it is just a cleaner way of implementing the interface... but maybe I'm not understanding your question.

Reply:
Ok, thank you. Think I may have interpreted what you meant incorrectly previously.

Was this what you had in mind? See https://github.com/google/go-github/pull/2795/commits/5618f6f4a68cd1cea4f355cda1dca67367ac50d3